python/lib/core/dmod/core/events/router.py

```diff
@@ -0,0 +1,135 @@
+            handlers: Handlers for events that should be registered
+        """
+        if events is None:
+            events = dict()
+
+        self.__fail_on_missing_event: bool = fail_on_missing_event or False
+        self.__allow_errors = allow_errors or True
+        self.__events: typing.Dict[str, EventFunctionGroup] = dict()
+
+        for event_name, signature in events.items():
+            self.register_event(event_name, signature)
+
+        for event_name, event_handlers in handlers.items():
+            self.register_handler(event_name, event_handlers)
+
+    def register_event(self, event: str, signature: SIGNATURE) -> EventRouter:
+        """
+        Register a handler signature with an event
+
+        Args:
+            event: The name of the event to register
+            signature: The expected signature for functions that will handle the event
+
+        Returns:
+            The updated EventRouter
+        """
+        if event in self.__events:
+            return self
+
+        if isinstance(signature, typing.Callable):
+            signature: Signature = Signature.from_function(signature)
+        elif isinstance(signature, inspect.Signature):
+            signature: Signature = Signature.from_signature(signature)
+        elif isinstance(signature, typing.Sequence):
+            signature = Signature(signature)
+
+        self.__events[event] = EventFunctionGroup(signature, allow_errors=self.__allow_errors)
+        return self
+
+    def register_handler(
+        self,
+        event: str,
+        handler: typing.Union[typing.Callable, typing.Sequence[typing.Callable]]
+    ) -> EventRouter:
+        if not handler:
+            return self
+
+        if not isinstance(handler, typing.Sequence):
+            handler = list(handler) if isinstance(handler, typing.Iterable) else [handler]
+
+        if not handler:
+            return self
+
+        if event not in self.__events:
+            logging.debug(
+                f"There is no registered event for '{event}' - "
+                f"a new event is being registered but the required signatures may be incorrect."
+            )
+            self.register_event(event, handler[0])
+
+        invalid_functions = list()
+
+        for event_handler in handler:
+            self.__events[event].add_function(event_handler, invalid_functions, allow_errors=self.__allow_errors)
+
+        if len(invalid_functions) > 0:
+            raise ValueError(
+                f"The following handlers cannot be added for the '{event}' event: {', '.join(invalid_functions)}"
+            )
+
+        return self
+
+    def __call__(self, event: typing.Union[str, Event], *args, **kwargs) -> typing.List[typing.Awaitable]:
+        if isinstance(event, str):
+            event = Event(event_name=event)
+
+        if not (self.__fail_on_missing_event or event.event_name in self.__events):
```

Comment:
This is a personal lint, so of course, feel free to ignore it. You could do something similar in `fire`.

```suggestion
        try:
            fn_grp = self.__events[event.event_name]
        except KeyError:
            if self.__fail_on_missing_event:
                raise ValueError(f"There are no registered handlers for the '{event.event_name}' event")
            return list()
        return fn_grp(event, *args, **kwargs)
```

Reply:
try/except has zero cost _if_ no exceptions are hit, but exceptions become costly if they _are_ hit. In this case, this might get hit _tons_ of times but never actually bring back functions, so instead of bypassing them via an if statement, this approach would instead trigger and catch loads of exceptions which we want to avoid. In this case, the logic was written based on most likely scenarios - that there aren't any handlers for a specific type of event (say we have tons of handlers for an `update` event but none for a `remove` event).

Reply:
reverse review - no acronyms!

I also found out a month or two ago that `variable = []` is a fair bit faster than `variable = list()`. I prefer seeing the constructor name, but `[]` is apparently the way python wants you to do it. Has something to do with function overhead and python's ability to copy the `[]` constant object. Only noticeable if you're going to perform the operation millions upon millions of times, though.